app/src/main/java/tech/relaycorp/letro/awala/processor/AwalaMessageProcessor.kt

```diff
@@ -3,17 +3,37 @@ package tech.relaycorp.letro.awala.processor
 import tech.relaycorp.awaladroid.messaging.IncomingMessage
 import tech.relaycorp.letro.awala.AwalaManager
 import tech.relaycorp.letro.awala.message.MessageType
+import tech.relaycorp.letro.utils.Logger
 
 interface AwalaMessageProcessor {
+    suspend fun isFromExpectedSender(
+        message: IncomingMessage,
+        awalaManager: AwalaManager,
+    ): Boolean
     suspend fun process(message: IncomingMessage, awalaManager: AwalaManager)
 }
 
 class AwalaMessageProcessorImpl constructor(
     private val processors: Map<MessageType, AwalaMessageProcessor>,
+    private val logger: Logger,
 ) : AwalaMessageProcessor {
 
     override suspend fun process(message: IncomingMessage, awalaManager: AwalaManager) {
         val type = MessageType.from(message.type)
-        processors[type]!!.process(message, awalaManager)
+        val processor = processors[type]!!
+        if (processor.isFromExpectedSender(message, awalaManager)) {
+            processors[type]!!.process(message, awalaManager)
+        } else {
+            logger.w(TAG, IllegalStateException("There is a message processor to process the message $type, but it came from unexpected sender"))
+        }
+    }
+
+    override suspend fun isFromExpectedSender(
+        message: IncomingMessage,
+        awalaManager: AwalaManager,
+    ): Boolean {
+        return true
```

Comment:
This makes it possible to write an unsafe processor, so I'd suggest to make this class abstract and not define this method so that subclasses are forced to define it.

Reply:
This is the actual implementation `AwalaMessageProcessorImpl` of `AwalaMessageProcessor` interface. It's a final class, which delegates processing to other processors. So I don't see how it's possible to write an unsafe processor, because this class is final. But other processors are implementing `AwalaMessageProcessor` interface, and must override 
```
abstract suspend fun isFromExpectedSender(
        message: IncomingMessage,
        awalaManager: AwalaManager,
    ): Boolean
```

Reply:
I see! But now I'm confused: What's calling `AwalaMessageProcessorImpl.isFromExpectedSender()` then?